src/sync/sync_user.cpp

```diff
@@ -0,0 +1,166 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync_user.hpp"
+
+#include "impl/sync_metadata.hpp"
+#include "sync_manager.hpp"
+#include "sync_session.hpp"
+
+namespace realm {
+
+SyncUser::SyncUser(std::string refresh_token, std::string identity, bool is_admin)
+: m_state(State::Active)
+, m_is_admin(is_admin)
+, m_refresh_token(std::move(refresh_token))
+, m_identity(std::move(identity))
+{ }
+
+std::vector<std::shared_ptr<SyncSession>> SyncUser::all_sessions()
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    std::vector<std::string> keys_to_remove;
+    std::vector<std::shared_ptr<SyncSession>> sessions;
+    if (m_state == State::Error) {
+        return sessions;
+    }
+    for (auto it : m_sessions) {
```

Comment:
This could use `std::remove_if()` to avoid building the temporary vector of keys and the hash table lookups to remove them:

```
auto it = std::remove_if(m_sessions.begin(), m_sessions.end(), [&](auto& session) {
    auto ptr = session.second.lock();
    if (ptr && ptr->is_valid()) {
        sessions.push_back(std::move(ptr));
        return false;
    }
    return true;
});
m_sessions.erase(it, m_sessions.end());
```